### PR TITLE
support pyramid images from OMERO

### PR DESCRIPTION
**Pull request by will-moore**
This allows me to open a large (96k x 45k) single Z/T tiled Image as an rgb image in napari.

Performance is not great.
And there are a LOT of calls to `get_tile()` (really needs some caching) and napari dies before
too long.
So this isn't ready to go, but may be a useful starting point.

### CODE REVIEW

**Comment by tlambert03** [comment]
great!  will take a look now and follow your updates.

> And there are a LOT of calls to get_tile() (really needs some caching) and napari dies before
too long.

that reminds me about one thing I wanted to mention about caching with dask arrays:  as of https://github.com/napari/napari/pull/1173, napari has support for [dask opportunistic caching ](https://docs.dask.org/en/latest/caching.html).  So if we provide a dask array as layer data, it will be cached automatically.  (That's why I removed the manual cache from the `get_lazy_data` function).  If you find any errors or performance issues with that, let me know.  but it _should_ just work.  And if you want to control the cache size, use `napari.utils.resize_dask_cache(nbytes)`

**Comment by joshmoore** [comment]
@will-moore : could you give something like this a try:

```
diff --git a/src/napari_omero/plugins/loaders.py b/src/napari_omero/plugins/loaders.py
index 4eeef0a..86c47be 100644
--- a/src/napari_omero/plugins/loaders.py
+++ b/src/napari_omero/plugins/loaders.py
@@ -159,6 +159,7 @@ def get_pyramid_lazy(image: ImageWrapper) -> List[da.Array]:
     """Get a pyramid of rgb dask arrays, loading tiles from OMERO."""
     size_z = image.getSizeZ()
     size_t = image.getSizeT()
+    size_c = image.getSizeC()
     pixels = image.getPrimaryPixels()
     dtype = PIXEL_TYPES.get(pixels.getPixelsType().value, None)

@@ -175,12 +176,22 @@ def get_pyramid_lazy(image: ImageWrapper) -> List[da.Array]:
         # create a new ImageWrapper (and rendering engine) each time
         # to avoid sharing re state between processes
         temp_img = image._conn.getObject('Image', image.id)
-        tile = temp_img.renderJpegRegion(z, t, x, y, w, h, level, compression=1)
-        temp_img._re.close()
-        i = BytesIO(tile)
-        rv = Image.open(i)
-        # i.close()
-        tile_data = np.asarray(rv)
+
+        pix = image._conn.c.sf.createRawPixelsStore()
+        try:
+            pix.setPixelsId(temp_img.id, False)
+            pix.setResolutionLevel(level)
+            tiles = []
+            for c in range(size_c):
+                tiles.append(pix.getTile(z, c, t, x, y, w, h))
+        finally:
+            pix.close()
+
+        for c, tile in enumerate(tiles):
+            tile = np.frombuffer(tile, dtype=np.uint8)
+            tile = tile.reshape((h, w))
+            tiles[c] = tile
+        tile_data = np.dstack(tiles)
         tile_cache[tile_name] = tile_data
         return tile_data
```

(I'm hesitant to open a PR since I'm working from a merge of this branch into origin/master

File: src/napari_omero/plugins/loaders.py [hunk 1]
```diff
@@ -13,6 +17,9 @@
 from ..utils import parse_omero_url, timer, lookup_obj, PIXEL_TYPES
 from ..widgets import QGateWay
 
+from dask.cache import Cache
+cache = Cache(2e9)  # Leverage two gigabytes of memory
+cache.register()
```
Comment (tlambert03) [inline src/napari_omero/plugins/loaders.py:22]: from the commit message, looks like you might just be adding this intentionally for testing.  But just wanted to point out that napari also registers a global cache [here](https://github.com/napari/napari/blob/823b83e39fe89dc30c1002fba7a36160430afbd3/napari/utils/dask_utils.py#L14-L45).  So when `viewer.add_image(data)` eventually gets called, if `data` is a dask array, caching is turned on (starting [here](https://github.com/napari/napari/blob/823b83e39fe89dc30c1002fba7a36160430afbd3/napari/layers/base/base.py#L140))

**Comment by will-moore** [comment]
Thanks @joshmoore that's looking good.
So that means I don't *have* to return `rgb` images. It feels like it's more efficient for pathology rgb images where you don't want to toggle channels etc but I wonder if it'll cause problems when trying to work out which dims are which when saving ROIs etc as I was finding before.

Also, does it save us much to only need 1 delayed tile per region, instead of 1 tile per channel?
@tlambert03 how expensive are dask tiles?
It seems to take a while to create all the dask delayed tiles I need for a big image.

I'm testing with https://downloads.openmicroscopy.org/images/SVS/77917.svs (96999 x 45667) with tiles of 256 x 256.

I tried the dask cache https://github.com/tlambert03/napari-omero/pull/1/commits/7bbab38eedc682d9b3c6a03f8c18d4045b5b7fdd
but this doesn't seem to reduce the repeated calls to `get_tile()` for the same tile.
I added a basic manual cache to observe this, and that does give an improved user experience of zooming in to the image.

However, I still quickly get a crash from somewhere...

<details><summary> stack trace... </summary>


```
WARNING: Traceback (most recent call last):
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 825, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/_vispy/vispy_base_layer.py", line 246, in on_draw
    shape_threshold=self.node.canvas.size,
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/layers/base/base.py", line 696, in _update_multiscale
    self.refresh()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/layers/base/base.py", line 633, in refresh
    self.set_view_slice()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/layers/base/base.py", line 586, in set_view_slice
    self._set_view_slice()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/layers/image/image.py", line 556, in _set_view_slice
    self.events.scale()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/utils/event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/utils/event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/utils/event.py", line 523, in _invoke_callback
    cb(event)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/_vispy/vispy_base_layer.py", line 156, in _on_scale_change
    self.layer.corner_pixels = self.coordinates_of_canvas_corners()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/_vispy/vispy_base_layer.py", line 214, in coordinates_of_canvas_corners
    self._transform_position(self.node.canvas.size) + offset[::-1]
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/app/canvas.py", line 336, in size
    size = self._backend._vispy_get_size()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 434, in _vispy_get_size
    g = self.geometry()
RuntimeError: wrapped C/C++ object of type CanvasBackendDesktop has been deleted
14:22:10 WARNING Traceback (most recent call last):
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 825, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/_vispy/vispy_base_layer.py", line 246, in on_draw
    shape_threshold=self.node.canvas.size,
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/layers/base/base.py", line 696, in _update_multiscale
    self.refresh()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/layers/base/base.py", line 633, in refresh
    self.set_view_slice()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/layers/base/base.py", line 586, in set_view_slice
    self._set_view_slice()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/layers/image/image.py", line 556, in _set_view_slice
    self.events.scale()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/utils/event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/utils/event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/utils/event.py", line 523, in _invoke_callback
    cb(event)
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/_vispy/vispy_base_layer.py", line 156, in _on_scale_change
    self.layer.corner_pixels = self.coordinates_of_canvas_corners()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/napari/_vispy/vispy_base_layer.py", line 214, in coordinates_of_canvas_corners
    self._transform_position(self.node.canvas.size) + offset[::-1]
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/app/canvas.py", line 336, in size
    size = self._backend._vispy_get_size()
  File "/Users/wmoore/opt/miniconda3/envs/omero-napari/lib/python3.7/site-packages/vispy/app/backends/_qt.py", line 434, in _vispy_get_size
    g = self.geometry()
RuntimeError: wrapped C/C++ object of type CanvasBackendDesktop has been deleted
Abort trap: 6
```
</details>

Screenshot 2020-06-25 at 14 39 16

**Comment by tlambert03** [comment]
couple thoughts: 
- if the cache size (2GB) is smaller than the total dataset, and if the dataset is a single plane, then we might need to change the cache scoring algorithm, as it will just basically evict older values and thrash continuously if the cache isn't big enough.  We could potentially alter it to say, "once a chunk is in there, don't forget it...".

   > basic manual cache to observe this, and that does give an improved user experience of zooming in to the image.

   I can imagine this would be the case if your manual cache never evicts. but the dask cache is always thrashing.  One debugging tip is that you can look at the hashes stored in the dask cache at `cache.cache.heap.heap`

- the biggest problem I foresee here is on the napari side.  I need to look into this a bit, but I think that when the `_set_view_slice` gets called, it's going to request the _full_ plane of data.  Not just the tiles that are required for the current zoom level.  the actual renderer (vispy) is efficient... but I think we're still calling `asarray()` on the full 2D plane... so while multiresolution/pyramid data works, I don't believe that we're taking advantage of 2D chunked/tiled data yet (I could be wrong... but that's my recollection).

**Comment by will-moore** [comment]
Hmm - that's a shame. I seem to remember the pyramid behaviour of napari being "OK" when using remote zarr data (e.g. the covid image). At least not requesting a ton of tiles that weren't in the viewport.
But yes, it does seem in this case that more tiles than needed are being loaded.

**Comment by tlambert03** [comment]
well, as mentioned.  I might be way off base here.  I haven't been handling any pyramid stuff personally.  @sofroniewn can you shed light on this?  what is napari able to do with 2D tiles here?  (see my second comment in https://github.com/tlambert03/napari-omero/pull/1#issuecomment-649562864)

**Comment by sofroniewn** [comment]
Hi @tlambert03 @will-moore we did make some changes recently to the pyramid handling that may have resulted in a degradation of performance.

> but I think that when the _set_view_slice gets called, it's going to request the full plane of data.... not just the tiles that are required for the current zoom level.

This is not true, otherwise the enormous pyramid examples wouldn't even work locally (which they do for me). What's happening now is that we request data every time the camera moves (this is because we decided to omit this rounding parameter @jni were going back and forth on) which is causing the slow down, but we only request exactly what is needed. Before we requested a little bit extra padding around each tile so you could pan and zoom a little before having to make another request.

I could try adding that back in, but @pwinston is going to be working on large multiscale 2D data after https://github.com/napari/napari/pull/1354 which should yield significant performance improvements and a much better architecture overall, so I'm inclined to just wait until then. Having more concrete use cases like this ready to go for him to verify will be very helpful too.

**Comment by tlambert03** [comment]
> but I think that when the _set_view_slice gets called, it's going to request the full plane of data.... not just the tiles that are required for the current zoom level.

>> This is not true, otherwise the enormous pyramid examples wouldn't even work locally (which they do for me)

yeah ... I was coming to that conclusion myself! 😂 (and only-too-late remembering some previous conversations). thanks for your input!!

**Comment by pwinston** [comment]
I almost replied here and then realized this is a different repo!

I added a comment to 1354 on napari to kind of restate things in light of the above. Nothing new but just making sure I can spell it out:
https://github.com/napari/napari/pull/1354#issuecomment-649678242

**Comment by will-moore** [comment]
Closing for now...

**Pull request closed without merging**